subprojects/gdk_rust/gdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,1212 @@
+                }
+            }))
+        } else {
+            None
+        };
+
+        let checker = match self.network.id() {
+            NetworkId::Bitcoin(network) => {
+                let mut path: PathBuf = self.data_root.as_str().into();
+                path.push(format!("headers_chain_{}", network));
+                ChainOrVerifier::Chain(HeadersChain::new(path, network)?)
+            }
+            NetworkId::Elements(network) => {
+                let verifier = Verifier::new(network);
+                ChainOrVerifier::Verifier(verifier)
+            }
+        };
+
+        let mut headers_kind = match &self.url {
+            ElectrumUrl::Tls(url, validate) => {
+                let client = electrum_client::Client::new_ssl(url.as_str(), *validate)?;
+                HeadersKind::Tls(
+                    Headers::new(db.clone(), checker, client)?,
+                    url.to_string(),
+                    *validate,
+                )
+            }
+            ElectrumUrl::Plaintext(url) => {
+                let client = electrum_client::Client::new(&url)?;
+                HeadersKind::Plain(Headers::new(db.clone(), checker, client)?, url.to_string())
+            }
+        };
+        let (close_headers, r) = channel();
+        self.closer.senders.push(close_headers);
+        let mut chunk_size = DIFFCHANGE_INTERVAL as usize;
+        let headers_handle = thread::spawn(move || {
+            info!("starting headers thread");
+            'outer: loop {
+                if wait_or_close(&r, sync_interval) {
+                    info!("closing headers thread");
+                    break;
+                }
+
+                loop {
+                    match headers_kind.ask(chunk_size) {
+                        Ok(headers_found) => {
+                            if headers_found == 0 {
+                                chunk_size = 1
+                            } else {
+                                info!("headers found: {}", headers_found);
+                            }
+                        }
+                        Err(Error::InvalidHeaders) => {
+                            // this should handle reorgs and also broke IO writes update
+                            if headers_kind.remove(144).is_err() {
+                                break;
+                            }
+                        }
+                        Err(e) => {
+                            // usual error is because I reached the tip, trying asking half
+                            info!("error while asking headers {}", e);
+                            if chunk_size > 1 {
+                                chunk_size /= 2
+                            }
+                        }
+                    };
+                    if r.try_recv().is_ok() {
+                        info!("closing headers thread");
+                        break 'outer;
+                    }
+                    if chunk_size == 1 {
```

Comment:
We could reach this even if the headers weren't fetched correctly. And this will not only trigger when `chunk_size == 1` because we've already synced up to the tip and only interested in the next block, but possibly also due to the `/= 2` division. Could this be problematic?

Maybe add a `continue` in the error branches of the match so it won't reach this point? This also makes sense for the re-org case, so the client will try re-syncing from 144 blocks back immediately.

Reply:
restructured a bit in 30f046be9b952028be1cf49c2e6947d79a7fda5a